// warp-arbundles/build/node/cjs/src/signing/chains/ArweaveSigner.js
"use strict";
var __awaiter = (this && this.__awaiter) || function (thisArg, _arguments, P, generator) {
    function adopt(value) { return value instanceof P ? value : new P(function (resolve) { resolve(value); }); }
    return new (P || (P = Promise))(function (resolve, reject) {
        function fulfilled(value) { try { step(generator.next(value)); } catch (e) { reject(e); } }
        function rejected(value) { try { step(generator["throw"](value)); } catch (e) { reject(e); } }
        function step(result) { result.done ? resolve(result.value) : adopt(result.value).then(fulfilled, rejected); }
        step((generator = generator.apply(thisArg, _arguments || [])).next());
    });
};
var __importDefault = (this && this.__importDefault) || function (mod) {
    return (mod && mod.__esModule) ? mod : { "default": mod };
};
Object.defineProperty(exports, "__esModule", { value: true });
const base64url_1 = __importDefault(require("base64url"));
const utils_1 = require("../../nodeUtils.js");
const constants_1 = require("../../constants");
class ArweaveSigner {
    constructor(jwk) {
        this.signatureType = 1;
        this.ownerLength = constants_1.SIG_CONFIG[1].pubLength;
        this.signatureLength = constants_1.SIG_CONFIG[1].sigLength;
        this.pk = jwk.n;
        this.jwk = jwk;
    }
    get publicKey() {
        return base64url_1.default.toBuffer(this.pk);
    }
    sign(message) {
        return (0, utils_1.getCryptoDriver)().sign(this.jwk, message);
    }
    static verify(pk, message, signature) {
        return __awaiter(this, void 0, void 0, function* () {
            return yield (0, utils_1.getCryptoDriver)().verify(pk, message, signature);
        });
    }
}
exports.default = ArweaveSigner;
