// arlocal/bin/routes/blocks.js
"use strict";
var __awaiter = (this && this.__awaiter) || function (thisArg, _arguments, P, generator) {
    function adopt(value) { return value instanceof P ? value : new P(function (resolve) { resolve(value); }); }
    return new (P || (P = Promise))(function (resolve, reject) {
        function fulfilled(value) { try { step(generator.next(value)); } catch (e) { reject(e); } }
        function rejected(value) { try { step(generator["throw"](value)); } catch (e) { reject(e); } }
        function step(result) { result.done ? resolve(result.value) : adopt(result.value).then(fulfilled, rejected); }
        step((generator = generator.apply(thisArg, _arguments || [])).next());
    });
};
Object.defineProperty(exports, "__esModule", { value: true });
exports.blocksRouteViaHeight = exports.blocksRoute = void 0;
const block_1 = require("../db/block");
let blockDB;
function blocksRoute(ctx) {
    return __awaiter(this, void 0, void 0, function* () {
        try {
            if (!blockDB) {
                blockDB = new block_1.BlockDB(ctx.connection);
            }
            const indepHash = ctx.params.indep_hash;
            const { id, mined_at: timestamp, previous_block, txs, height } = yield blockDB.getByIndepHash(indepHash);
            ctx.body = {
                indep_hash: id,
                timestamp: Math.round(new Date(timestamp).getTime() / 1000),
                previous_block,
                // return block height instead of current height
                height,
                txs: txs.split(','),
            };
        }
        catch (error) {
            console.error({ error });
        }
    });
}
exports.blocksRoute = blocksRoute;
function blocksRouteViaHeight(ctx) {
    return __awaiter(this, void 0, void 0, function* () {
        try {
            if (!blockDB) {
                blockDB = new block_1.BlockDB(ctx.connection);
            }
            const h = parseInt(ctx.params.height, 10) || 0;
            const block = yield blockDB.getByHeight(h);
            if (h === 0 && !block) {
                // return genesis block
                // hash = genesis block hash
                let hash = ctx.network.current;
                if (ctx.network.height > 0) {
                    // find the previous block of the height 1 block
                    const blk = yield blockDB.getByHeight(1);
                    ({ previous_block: hash } = blk);
                }
                ctx.body = {
                    indep_hash: hash,
                    timestamp: Math.round(ctx.timestamp / 1000),
                    previous_block: '',
                    height: h,
                    txs: [''],
                };
                return;
            }
            const { id, mined_at: timestamp, previous_block, txs, height } = block;
            ctx.body = {
                indep_hash: id,
                timestamp: Math.round(new Date(timestamp).getTime() / 1000),
                previous_block,
                height,
                txs: txs.split(','),
            };
        }
        catch (error) {
            console.error({ error });
        }
    });
}
exports.blocksRouteViaHeight = blocksRouteViaHeight;
