// arlocal/bin/app.js
"use strict";
var __createBinding = (this && this.__createBinding) || (Object.create ? (function(o, m, k, k2) {
    if (k2 === undefined) k2 = k;
    Object.defineProperty(o, k2, { enumerable: true, get: function() { return m[k]; } });
}) : (function(o, m, k, k2) {
    if (k2 === undefined) k2 = k;
    o[k2] = m[k];
}));
var __setModuleDefault = (this && this.__setModuleDefault) || (Object.create ? (function(o, v) {
    Object.defineProperty(o, "default", { enumerable: true, value: v });
}) : function(o, v) {
    o["default"] = v;
});
var __importStar = (this && this.__importStar) || function (mod) {
    if (mod && mod.__esModule) return mod;
    var result = {};
    if (mod != null) for (var k in mod) if (k !== "default" && Object.prototype.hasOwnProperty.call(mod, k)) __createBinding(result, mod, k);
    __setModuleDefault(result, mod);
    return result;
};
var __awaiter = (this && this.__awaiter) || function (thisArg, _arguments, P, generator) {
    function adopt(value) { return value instanceof P ? value : new P(function (resolve) { resolve(value); }); }
    return new (P || (P = Promise))(function (resolve, reject) {
        function fulfilled(value) { try { step(generator.next(value)); } catch (e) { reject(e); } }
        function rejected(value) { try { step(generator["throw"](value)); } catch (e) { reject(e); } }
        function step(result) { result.done ? resolve(result.value) : adopt(result.value).then(fulfilled, rejected); }
        step((generator = generator.apply(thisArg, _arguments || [])).next());
    });
};
var __importDefault = (this && this.__importDefault) || function (mod) {
    return (mod && mod.__esModule) ? mod : { "default": mod };
};
Object.defineProperty(exports, "__esModule", { value: true });
const fs_1 = require("fs");
const path_1 = __importStar(require("path"));
const koa_1 = __importDefault(require("koa"));
const cors_1 = __importDefault(require("@koa/cors"));
const koa_bodyparser_1 = __importDefault(require("koa-bodyparser"));
const koa_router_1 = __importDefault(require("koa-router"));
const koa_json_1 = __importDefault(require("koa-json"));
const koa_logger_1 = __importDefault(require("koa-logger"));
const connect_1 = require("./db/connect");
const initialize_1 = require("./db/initialize");
const server_1 = require("./graphql/server");
const data_1 = require("./routes/data");
const mine_1 = require("./routes/mine");
const status_1 = require("./routes/status");
const transaction_1 = require("./routes/transaction");
const transaction_2 = require("./middlewares/transaction");
const utils_1 = require("./utils/utils");
const logging_1 = __importDefault(require("./utils/logging"));
const blocks_1 = require("./routes/blocks");
const wallet_1 = require("./routes/wallet");
const chunk_1 = require("./routes/chunk");
const peer_1 = require("./routes/peer");
const wallet_2 = require("./db/wallet");
const block_1 = require("./db/block");
class ArLocal {
    constructor(port = 1984, showLogs = true, dbPath, persist = false, fails = 0) {
        this.port = 1984;
        this.app = new koa_1.default();
        this.router = new koa_router_1.default();
        this.port = port || this.port;
        dbPath = dbPath || path_1.default.join(__dirname, '.db', port.toString());
        this.dbPath = dbPath;
        this.persist = persist;
        this.fails = fails;
        this.log = new logging_1.default(showLogs);
        this.connection = (0, connect_1.connect)(dbPath);
        this.app.context.network = {
            network: 'arlocal.N.1',
            version: 1,
            release: 1,
            queue_length: 0,
            peers: 1,
            height: 0,
            current: utils_1.Utils.randomID(64),
            blocks: 1,
            node_state_latency: 0,
        };
        this.app.context.logging = this.log;
        this.app.context.dbPath = dbPath;
        this.app.context.connection = this.connection;
        this.app.context.fails = this.fails / 100;
        // server start date for genesis block timestamp
        this.app.context.timestamp = new Date().getTime();
        this.walletDB = new wallet_2.WalletDB(this.connection);
    }
    start() {
        return __awaiter(this, void 0, void 0, function* () {
            yield this.startDB();
            const blockDB = new block_1.BlockDB(this.connection);
            const lastBlock = yield blockDB.getLastBlock();
            if (lastBlock) {
                this.app.context.network.current = lastBlock.id;
                this.app.context.network.height = lastBlock.height;
                this.app.context.network.blocks = lastBlock.height + 1;
            }
            else {
                // save the genesis block to db
                yield blockDB.insertGenesis(this.app.context.network.current);
            }
            // ISSUE WITH ARCONNECT THAT IS SENDING HEAD REQUESTS FOR NO REASON
            this.router.use((ctx, next) => __awaiter(this, void 0, void 0, function* () {
                if (ctx.method === 'HEAD') {
                    ctx.status = 405;
                    ctx.body = 'Method Not Allowed';
                    return;
                }
                yield next();
            }));
            this.router.get('/', status_1.statusRoute);
            this.router.get('/info', status_1.statusRoute);
            this.router.get('/peers', peer_1.peersRoute);
            this.router.get('/mine/:qty?', mine_1.mineRoute);
            this.router.get('/mineWithFails/:qty?', mine_1.mineWithFailsRoute);
            this.router.get('/tx_anchor', transaction_1.txAnchorRoute);
            this.router.get('/price/:bytes/:addy?', (ctx) => __awaiter(this, void 0, void 0, function* () { return (ctx.body = +ctx.params.bytes * 1965132); }));
            // tx filter endpoint to restrict ans-104 txs
            this.router.get(/^\/tx(?:\/|$)/, transaction_2.txAccessMiddleware);
            this.router.get('/tx/:txid/offset', transaction_1.txOffsetRoute);
            this.router.get('/tx/:txid/status', transaction_1.txStatusRoute);
            this.router.get('/tx/:txid/data', transaction_1.txRawDataRoute);
            this.router.get('/tx/:txid/data.:ext', transaction_1.txDataRoute);
            this.router.get('/tx/:txid/:field', transaction_1.txFieldRoute);
            this.router.get('/tx/:txid/:file', transaction_1.txFileRoute);
            this.router.get('/tx/:txid', transaction_1.txRoute);
            this.router.delete('/tx/:txid', transaction_1.deleteTxRoute);
            this.router.post('/tx', transaction_1.txPostRoute);
            this.router.post('/chunk', chunk_1.postChunkRoute);
            this.router.get('/chunk/:offset', chunk_1.getChunkOffsetRoute);
            this.router.get('/block/hash/:indep_hash', blocks_1.blocksRoute);
            this.router.get('/block/height/:height', blocks_1.blocksRouteViaHeight);
            this.router.post('/wallet', wallet_1.createWalletRoute);
            this.router.patch('/wallet/:address/balance', wallet_1.updateBalanceRoute);
            this.router.get('/mint/:address/:balance', wallet_1.addBalanceRoute);
            this.router.get('/wallet/:address/balance', wallet_1.getBalanceRoute);
            this.router.get('/wallet/:address/last_tx', wallet_1.getLastWalletTxRoute);
            this.router.head(data_1.dataRouteRegex, data_1.dataHeadRoute);
            this.router.get(data_1.dataRouteRegex, data_1.dataRoute);
            this.router.get('/(.*)', data_1.subDataRoute);
            this.router.get('/:other', (ctx) => {
                ctx.type = 'application/json';
                ctx.body = {
                    status: 400,
                    error: 'Request type not found.',
                };
            });
            this.app.use((0, cors_1.default)({
                origin: '*',
            }));
            this.app.use((ctx, next) => __awaiter(this, void 0, void 0, function* () {
                // set cross origin resource policy header
                try {
                    ctx.set('Cross-Origin-Resource-Policy', 'cross-origin');
                }
                catch (_a) { }
                return yield next();
            }));
            this.app.use((0, koa_json_1.default)());
            this.app.use((0, koa_logger_1.default)({
                transporter: (str) => {
                    this.log.log(str);
                },
            }));
            this.app.use((0, koa_bodyparser_1.default)());
            this.app.use(this.router.routes()).use(this.router.allowedMethods());
            this.server = this.app.listen(this.port, () => {
                console.log(`arlocal started on port ${this.port}`);
            });
        });
    }
    startDB() {
        return __awaiter(this, void 0, void 0, function* () {
            // Delete old database
            try {
                if (!this.persist)
                    (0, fs_1.rmSync)(this.dbPath, { recursive: true });
            }
            catch (e) { }
            if (!(0, fs_1.existsSync)(this.dbPath))
                (0, fs_1.mkdirSync)(this.dbPath, { recursive: true });
            // sqlite
            this.apollo = (0, server_1.graphServer)({
                introspection: true,
                playground: true,
            }, this.connection);
            yield this.apollo.start();
            this.apollo.applyMiddleware({ app: this.app, path: '/graphql' });
            if (!(0, fs_1.existsSync)((0, path_1.join)(this.dbPath, 'db.sqlite')))
                yield (0, initialize_1.up)(this.connection);
        });
    }
    stop() {
        return __awaiter(this, void 0, void 0, function* () {
            if (this.server) {
                this.server.close((err) => {
                    if (err) {
                        try {
                            if (!this.persist)
                                (0, fs_1.rmSync)(this.dbPath, { recursive: true });
                        }
                        catch (err) { }
                        return;
                    }
                });
            }
            (0, initialize_1.down)(this.connection, this.persist)
                .then(() => {
                this.apollo
                    .stop()
                    .then(() => {
                    this.connection
                        .destroy()
                        .then(() => {
                        try {
                            if (!this.persist)
                                (0, fs_1.rmSync)(this.dbPath, { recursive: true });
                        }
                        catch (e) { }
                    })
                        .catch(() => { });
                })
                    .catch(() => { });
            })
                .catch(() => { });
        });
    }
    getServer() {
        return this.server;
    }
    getNetwork() {
        return this.app.context.network;
    }
    getDbPath() {
        return this.dbPath;
    }
    getWalletDb() {
        return this.walletDB;
    }
}
exports.default = ArLocal;
