// smartweave/lib/contract-load.js
"use strict";
var __createBinding = (this && this.__createBinding) || (Object.create ? (function(o, m, k, k2) {
    if (k2 === undefined) k2 = k;
    Object.defineProperty(o, k2, { enumerable: true, get: function() { return m[k]; } });
}) : (function(o, m, k, k2) {
    if (k2 === undefined) k2 = k;
    o[k2] = m[k];
}));
var __setModuleDefault = (this && this.__setModuleDefault) || (Object.create ? (function(o, v) {
    Object.defineProperty(o, "default", { enumerable: true, value: v });
}) : function(o, v) {
    o["default"] = v;
});
var __importStar = (this && this.__importStar) || function (mod) {
    if (mod && mod.__esModule) return mod;
    var result = {};
    if (mod != null) for (var k in mod) if (k !== "default" && Object.prototype.hasOwnProperty.call(mod, k)) __createBinding(result, mod, k);
    __setModuleDefault(result, mod);
    return result;
};
var __awaiter = (this && this.__awaiter) || function (thisArg, _arguments, P, generator) {
    function adopt(value) { return value instanceof P ? value : new P(function (resolve) { resolve(value); }); }
    return new (P || (P = Promise))(function (resolve, reject) {
        function fulfilled(value) { try { step(generator.next(value)); } catch (e) { reject(e); } }
        function rejected(value) { try { step(generator["throw"](value)); } catch (e) { reject(e); } }
        function step(result) { result.done ? resolve(result.value) : adopt(result.value).then(fulfilled, rejected); }
        step((generator = generator.apply(thisArg, _arguments || [])).next());
    });
};
var __importDefault = (this && this.__importDefault) || function (mod) {
    return (mod && mod.__esModule) ? mod : { "default": mod };
};
Object.defineProperty(exports, "__esModule", { value: true });
exports.createContractExecutionEnvironment = exports.loadContract = void 0;
const clarity = __importStar(require("@weavery/clarity"));
const utils_1 = require("./utils");
const smartweave_global_1 = require("./smartweave-global");
const bignumber_js_1 = __importDefault(require("bignumber.js"));
/**
 * Loads the contract source, initial state and other parameters
 *
 * @param arweave     an Arweave client instance
 * @param contractID  the Transaction Id of the contract
 */
function loadContract(arweave, contractID, contractSrcTXID) {
    return __awaiter(this, void 0, void 0, function* () {
        // Generate an object containing the details about a contract in one place.
        const contractTX = yield arweave.transactions.get(contractID);
        const contractOwner = yield arweave.wallets.ownerToAddress(contractTX.owner);
        contractSrcTXID = contractSrcTXID || utils_1.getTag(contractTX, 'Contract-Src');
        const minFee = utils_1.getTag(contractTX, 'Min-Fee');
        const contractSrcTX = yield arweave.transactions.get(contractSrcTXID);
        const contractSrc = contractSrcTX.get('data', { decode: true, string: true });
        let state;
        if (utils_1.getTag(contractTX, 'Init-State')) {
            state = utils_1.getTag(contractTX, 'Init-State');
        }
        else if (utils_1.getTag(contractTX, 'Init-State-TX')) {
            const stateTX = yield arweave.transactions.get(utils_1.getTag(contractTX, 'Init-State-TX'));
            state = stateTX.get('data', { decode: true, string: true });
        }
        else {
            state = contractTX.get('data', { decode: true, string: true });
        }
        const { handler, swGlobal } = createContractExecutionEnvironment(arweave, contractSrc, contractID, contractOwner);
        return {
            id: contractID,
            contractSrcTXID,
            contractSrc,
            initState: state,
            minFee,
            contractTX,
            handler,
            swGlobal,
        };
    });
}
exports.loadContract = loadContract;
/**
 * Translates a contract source code into a Js function that can be called, and sets
 * up two globals, SmartWeave and the ContractError exception.
 *
 * At the moment this uses the Function() constructor (basically the same as eval),
 * But the design is geared toward switching to Realms or something like
 * https://github.com/justjake/quickjs-emscripten. (probably the latter)
 *
 * In the current implemention, using Function(), the 'globals' are actually
 * just lexically scoped vars, unique to each instance of a contract.
 *
 * @param contractSrc the javascript source for the contract. Must declare a handle() function
 */
function createContractExecutionEnvironment(arweave, contractSrc, contractId, contractOwner) {
    const returningSrc = utils_1.normalizeContractSource(contractSrc);
    const swGlobal = new smartweave_global_1.SmartWeaveGlobal(arweave, { id: contractId, owner: contractOwner });
    const getContractFunction = new Function(returningSrc); // eslint-disable-line
    return {
        handler: getContractFunction(swGlobal, bignumber_js_1.default, clarity),
        swGlobal,
    };
}
exports.createContractExecutionEnvironment = createContractExecutionEnvironment;
