// smartweave/lib/contract-interact.js
"use strict";
var __awaiter = (this && this.__awaiter) || function (thisArg, _arguments, P, generator) {
    function adopt(value) { return value instanceof P ? value : new P(function (resolve) { resolve(value); }); }
    return new (P || (P = Promise))(function (resolve, reject) {
        function fulfilled(value) { try { step(generator.next(value)); } catch (e) { reject(e); } }
        function rejected(value) { try { step(generator["throw"](value)); } catch (e) { reject(e); } }
        function step(result) { result.done ? resolve(result.value) : adopt(result.value).then(fulfilled, rejected); }
        step((generator = generator.apply(thisArg, _arguments || [])).next());
    });
};
var __importDefault = (this && this.__importDefault) || function (mod) {
    return (mod && mod.__esModule) ? mod : { "default": mod };
};
Object.defineProperty(exports, "__esModule", { value: true });
exports.interactRead = exports.interactWriteDryRunCustom = exports.interactWriteDryRun = exports.simulateInteractWrite = exports.interactWrite = void 0;
const contract_load_1 = require("./contract-load");
const contract_read_1 = require("./contract-read");
const contract_step_1 = require("./contract-step");
const utils_1 = require("./utils");
const errors_1 = __importDefault(require("./errors"));
/**
 * Writes an interaction on the blockchain.
 *
 * This simply creates an interaction tx and posts it.
 * It does not need to know the current state of the contract.
 *
 * @param arweave       an Arweave client instance
 * @param wallet        a wallet private key
 * @param contractId    the Transaction Id of the contract
 * @param input         the interaction input, will be serialized as Json.
 * @param tags          an array of tags with name/value as objects.
 * @param target        if needed to send AR to an address, this is the target.
 * @param winstonQty    amount of winston to send to the target, if needed.
 * @param reward        custom reward for txs, if needed.
 */
function interactWrite(arweave, wallet, contractId, input, tags = [], target = '', winstonQty = '', reward) {
    return __awaiter(this, void 0, void 0, function* () {
        const interactionTx = yield createTx(arweave, wallet, contractId, input, tags, target, winstonQty, reward);
        const response = yield arweave.transactions.post(interactionTx);
        if (response.status !== 200)
            return null;
        return interactionTx.id;
    });
}
exports.interactWrite = interactWrite;
/**
 * Simulates an interaction on the blockchain and returns the simulated transaction.
 *
 * This simply creates an interaction tx and posts it.
 * It does not need to know the current state of the contract.
 *
 * @param arweave       an Arweave client instance
 * @param wallet        a wallet private key
 * @param contractId    the Transaction Id of the contract
 * @param input         the interaction input, will be serialized as Json.
 * @param tags          an array of tags with name/value as objects.
 * @param target        if needed to send AR to an address, this is the target.
 * @param winstonQty    amount of winston to send to the target, if needed.
 */
function simulateInteractWrite(arweave, wallet, contractId, input, tags = [], target = '', winstonQty = '') {
    return __awaiter(this, void 0, void 0, function* () {
        const interactionTx = yield createTx(arweave, wallet, contractId, input, tags, target, winstonQty);
        return interactionTx;
    });
}
exports.simulateInteractWrite = simulateInteractWrite;
/**
 * This will load a contract to its latest state, and do a dry run of an interaction,
 * without writing anything to the chain.
 *
 * @param arweave       an Arweave client instance
 * @param wallet        a wallet private or public key
 * @param contractId    the Transaction Id of the contract
 * @param input         the interaction input.
 * @param tags          an array of tags with name/value as objects.
 * @param target        if needed to send AR to an address, this is the target.
 * @param winstonQty    amount of winston to send to the target, if needed.
 * @param myState       a locally-generated state variable
 * @param fromParam     The from address of the transaction
 * @param contractInfoParam The loaded contract
 */
function interactWriteDryRun(arweave, wallet, contractId, input, tags = [], target = '', winstonQty = '', myState, fromParam, contractInfoParam) {
    return __awaiter(this, void 0, void 0, function* () {
        // tslint:disable-next-line: prefer-const
        let { handler, swGlobal, contractSrcTXID } = contractInfoParam || (yield contract_load_1.loadContract(arweave, contractId));
        const latestState = myState || (yield contract_read_1.readContract(arweave, contractId));
        const from = fromParam || (yield arweave.wallets.getAddress(wallet));
        const settings = utils_1.evalSettings(latestState);
        const evolve = latestState.evolve || settings.get('evolve');
        let canEvolve = latestState.canEvolve || settings.get('canEvolve');
        // By default, contracts can evolve if there's not an explicit `false`.
        if (canEvolve === undefined || canEvolve === null) {
            canEvolve = true;
        }
        if (evolve && /[a-z0-9_-]{43}/i.test(evolve) && canEvolve) {
            if (contractSrcTXID !== evolve) {
                try {
                    const contractInfo = yield contract_load_1.loadContract(arweave, contractId, evolve);
                    swGlobal = contractInfo.swGlobal;
                    handler = contractInfo.handler;
                    contractSrcTXID = evolve;
                }
                catch (e) {
                    const error = new errors_1.default("CONTRACT_NOT_FOUND" /* CONTRACT_NOT_FOUND */, {
                        message: `Contract having txId: ${contractId} not found`,
                        requestedTxId: contractId,
                    });
                    console.log(error);
                }
            }
        }
        const interaction = {
            input,
            caller: from,
        };
        const tx = yield createTx(arweave, wallet, contractId, input, tags, target, winstonQty);
        const ts = utils_1.unpackTags(tx);
        const currentBlock = yield arweave.blocks.getCurrent();
        swGlobal._activeTx = createDummyTx(tx, from, ts, currentBlock);
        return yield contract_step_1.execute(handler, interaction, latestState);
    });
}
exports.interactWriteDryRun = interactWriteDryRun;
/**
 * This will load a contract to its latest state, and do a dry run of an interaction,
 * without writing anything to the chain.
 *
 * @param arweave       an Arweave client instance
 * @param tx            a signed transaction
 * @param contractId    the Transaction Id of the contract
 * @param input         the interaction input.
 * @param myState       a locally-generated state variable
 * @param fromParam     The from address of the transaction
 * @param contractInfoParam The loaded contract
 */
function interactWriteDryRunCustom(arweave, tx, contractId, input, myState, fromParam = {}, contractInfoParam) {
    return __awaiter(this, void 0, void 0, function* () {
        // tslint:disable-next-line: prefer-const
        let { handler, swGlobal, contractSrcTXID } = contractInfoParam || (yield contract_load_1.loadContract(arweave, contractId));
        const latestState = myState || (yield contract_read_1.readContract(arweave, contractId));
        const from = fromParam;
        const settings = utils_1.evalSettings(latestState);
        const evolve = latestState.evolve || settings.get('evolve');
        let canEvolve = latestState.canEvolve || settings.get('canEvolve');
        // By default, contracts can evolve if there's not an explicit `false`.
        if (canEvolve === undefined || canEvolve === null) {
            canEvolve = true;
        }
        if (evolve && /[a-z0-9_-]{43}/i.test(evolve) && canEvolve) {
            if (contractSrcTXID !== evolve) {
                try {
                    const contractInfo = yield contract_load_1.loadContract(arweave, contractId, evolve);
                    swGlobal = contractInfo.swGlobal;
                    handler = contractInfo.handler;
                    contractSrcTXID = evolve;
                }
                catch (e) {
                    const error = new errors_1.default("CONTRACT_NOT_FOUND" /* CONTRACT_NOT_FOUND */, {
                        message: `Contract having txId: ${contractId} not found`,
                        requestedTxId: contractId,
                    });
                    console.log(error);
                }
            }
        }
        const interaction = {
            input,
            caller: from,
        };
        const ts = utils_1.unpackTags(tx);
        const currentBlock = yield arweave.blocks.getCurrent();
        swGlobal._activeTx = createDummyTx(tx, from, ts, currentBlock);
        return yield contract_step_1.execute(handler, interaction, latestState);
    });
}
exports.interactWriteDryRunCustom = interactWriteDryRunCustom;
/**
 * This will load a contract to its latest state, and execute a read interaction that
 * does not change any state.
 *
 * @param arweave       an Arweave client instance
 * @param wallet        a wallet private or public key
 * @param contractId    the Transaction Id of the contract
 * @param input         the interaction input.
 * @param tags          an array of tags with name/value as objects.
 * @param target        if needed to send AR to an address, this is the target.
 * @param winstonQty    amount of winston to send to the target, if needed.
 */
function interactRead(arweave, wallet, contractId, input, tags = [], target = '', winstonQty = '') {
    return __awaiter(this, void 0, void 0, function* () {
        // tslint:disable-next-line: prefer-const
        let { handler, swGlobal, contractSrcTXID } = yield contract_load_1.loadContract(arweave, contractId);
        const latestState = yield contract_read_1.readContract(arweave, contractId);
        const from = wallet ? yield arweave.wallets.getAddress(wallet) : '';
        const settings = utils_1.evalSettings(latestState);
        const evolve = latestState.evolve || settings.get('evolve');
        let canEvolve = latestState.canEvolve || settings.get('canEvolve');
        // By default, contracts can evolve if there's not an explicit `false`.
        if (canEvolve === undefined || canEvolve === null) {
            canEvolve = true;
        }
        if (evolve && /[a-z0-9_-]{43}/i.test(evolve) && canEvolve) {
            if (contractSrcTXID !== evolve) {
                try {
                    const contractInfo = yield contract_load_1.loadContract(arweave, contractId, evolve);
                    swGlobal = contractInfo.swGlobal;
                    handler = contractInfo.handler;
                    contractSrcTXID = evolve;
                }
                catch (e) {
                    const error = new errors_1.default("CONTRACT_NOT_FOUND" /* CONTRACT_NOT_FOUND */, {
                        message: `Contract having txId: ${contractId} not found`,
                        requestedTxId: contractId,
                    });
                    console.log(error);
                }
            }
        }
        const interaction = {
            input,
            caller: from,
        };
        const tx = yield createTx(arweave, wallet, contractId, input, tags, target, winstonQty);
        const ts = utils_1.unpackTags(tx);
        const currentBlock = yield arweave.blocks.getCurrent();
        swGlobal._activeTx = createDummyTx(tx, from, ts, currentBlock);
        const result = yield contract_step_1.execute(handler, interaction, latestState);
        return result.result;
    });
}
exports.interactRead = interactRead;
function createTx(arweave, wallet, contractId, input, tags, target = '', winstonQty = '0', reward) {
    return __awaiter(this, void 0, void 0, function* () {
        const options = {
            data: Math.random().toString().slice(-4),
            reward,
        };
        if (target && target.length) {
            options.target = target.toString();
            if (winstonQty && +winstonQty > 0) {
                options.quantity = winstonQty.toString();
            }
        }
        const interactionTx = yield arweave.createTransaction(options, wallet);
        if (!input) {
            throw new Error(`Input should be a truthy value: ${JSON.stringify(input)}`);
        }
        if (tags && tags.length) {
            for (const tag of tags) {
                interactionTx.addTag(tag.name.toString(), tag.value.toString());
            }
        }
        interactionTx.addTag('App-Name', 'SmartWeaveAction');
        interactionTx.addTag('App-Version', '0.3.0');
        interactionTx.addTag('Contract', contractId);
        interactionTx.addTag('Input', JSON.stringify(input));
        yield arweave.transactions.sign(interactionTx, wallet);
        return interactionTx;
    });
}
function createDummyTx(tx, from, tags, block) {
    return {
        id: tx.id,
        owner: {
            address: from,
        },
        recipient: tx.target,
        tags,
        fee: {
            winston: tx.reward,
        },
        quantity: {
            winston: tx.quantity,
        },
        block: {
            id: block.indep_hash,
            height: block.height,
            timestamp: block.timestamp,
        },
    };
}
