// smartweave/lib/contract-create.js
"use strict";
var __awaiter = (this && this.__awaiter) || function (thisArg, _arguments, P, generator) {
    function adopt(value) { return value instanceof P ? value : new P(function (resolve) { resolve(value); }); }
    return new (P || (P = Promise))(function (resolve, reject) {
        function fulfilled(value) { try { step(generator.next(value)); } catch (e) { reject(e); } }
        function rejected(value) { try { step(generator["throw"](value)); } catch (e) { reject(e); } }
        function step(result) { result.done ? resolve(result.value) : adopt(result.value).then(fulfilled, rejected); }
        step((generator = generator.apply(thisArg, _arguments || [])).next());
    });
};
Object.defineProperty(exports, "__esModule", { value: true });
exports.createContractFromTx = exports.createContract = exports.simulateCreateContractFromTx = exports.simulateCreateContractFromSource = void 0;
/**
 * Simulates the creation of a new contract from a contract, so that the cost for the creation can be checked
 * Returns the transaction describing the creation simulation.
 *
 * @param arweave       an Arweave client instance
 * @param wallet        a wallet private or public key
 * @param initState     the contract initial state, as a JSON string.
 * @param contractSrc optional the contract source as string.
 */
function simulateCreateContractFromSource(arweave, wallet, initState, contractSrc, reward) {
    return __awaiter(this, void 0, void 0, function* () {
        const srcTx = yield arweave.createTransaction({ data: contractSrc, reward }, wallet);
        srcTx.addTag('App-Name', 'SmartWeaveContractSource');
        srcTx.addTag('App-Version', '0.3.0');
        srcTx.addTag('Content-Type', 'application/javascript');
        yield arweave.transactions.sign(srcTx, wallet);
        // compute the fee needed to deploy the init state
        const deployInitStateTx = yield simulateCreateContractFromTx(arweave, wallet, srcTx.id, initState);
        const initStateReward = deployInitStateTx.reward;
        // update the reward of the contract creation by adding the reward needed for the creation of the state
        srcTx.reward = (parseFloat(srcTx.reward) + parseFloat(initStateReward)).toString();
        return srcTx;
    });
}
exports.simulateCreateContractFromSource = simulateCreateContractFromSource;
/**
 * Simulate the creation of a contract from an existing contract source tx, with an initial state.
 * Returns the contract id.
 *
 * @param arweave   an Arweave client instance
 * @param wallet    a wallet private or public key
 * @param srcTxId   the contract source Tx id.
 * @param state     the initial state, as a JSON string.
 * @param tags          an array of tags with name/value as objects.
 * @param target        if needed to send AR to an address, this is the target.
 * @param winstonQty    amount of winston to send to the target, if needed.
 */
function simulateCreateContractFromTx(arweave, wallet, srcTxId, state, tags = [], target = '', winstonQty = '', reward) {
    return __awaiter(this, void 0, void 0, function* () {
        let contractTX = yield arweave.createTransaction({ data: state, reward }, wallet);
        if (target && winstonQty && target.length && +winstonQty > 0) {
            contractTX = yield arweave.createTransaction({
                data: state,
                target: target.toString(),
                quantity: winstonQty.toString(),
                reward,
            }, wallet);
        }
        if (tags && tags.length) {
            for (const tag of tags) {
                contractTX.addTag(tag.name.toString(), tag.value.toString());
            }
        }
        contractTX.addTag('App-Name', 'SmartWeaveContract');
        contractTX.addTag('App-Version', '0.3.0');
        contractTX.addTag('Contract-Src', srcTxId);
        contractTX.addTag('Content-Type', 'application/json');
        yield arweave.transactions.sign(contractTX, wallet);
        return contractTX;
    });
}
exports.simulateCreateContractFromTx = simulateCreateContractFromTx;
/**
 * Create a new contract from a contract source file and an initial state.
 * Returns the contract id.
 *
 * @param arweave       an Arweave client instance
 * @param wallet        a wallet private or public key
 * @param contractSrc   the contract source as string.
 * @param initState     the contract initial state, as a JSON string.
 */
function createContract(arweave, wallet, contractSrc, initState, reward) {
    return __awaiter(this, void 0, void 0, function* () {
        const srcTx = yield arweave.createTransaction({ data: contractSrc, reward }, wallet);
        srcTx.addTag('App-Name', 'SmartWeaveContractSource');
        srcTx.addTag('App-Version', '0.3.0');
        srcTx.addTag('Content-Type', 'application/javascript');
        yield arweave.transactions.sign(srcTx, wallet);
        const response = yield arweave.transactions.post(srcTx);
        if (response.status === 200 || response.status === 208) {
            return yield createContractFromTx(arweave, wallet, srcTx.id, initState);
        }
        else {
            throw new Error('Unable to write Contract Source.');
        }
    });
}
exports.createContract = createContract;
/**
 * Create a new contract from an existing contract source tx, with an initial state.
 * Returns the contract id.
 *
 * @param arweave   an Arweave client instance
 * @param wallet    a wallet private or public key
 * @param srcTxId   the contract source Tx id.
 * @param state     the initial state, as a JSON string.
 * @param tags          an array of tags with name/value as objects.
 * @param target        if needed to send AR to an address, this is the target.
 * @param winstonQty    amount of winston to send to the target, if needed.
 */
function createContractFromTx(arweave, wallet, srcTxId, state, tags = [], target = '', winstonQty = '', reward) {
    return __awaiter(this, void 0, void 0, function* () {
        let contractTX = yield arweave.createTransaction({ data: state, reward }, wallet);
        if (target && winstonQty && target.length && +winstonQty > 0) {
            contractTX = yield arweave.createTransaction({
                data: state,
                target: target.toString(),
                quantity: winstonQty.toString(),
                reward,
            }, wallet);
        }
        if (tags && tags.length) {
            for (const tag of tags) {
                contractTX.addTag(tag.name.toString(), tag.value.toString());
            }
        }
        contractTX.addTag('App-Name', 'SmartWeaveContract');
        contractTX.addTag('App-Version', '0.3.0');
        contractTX.addTag('Contract-Src', srcTxId);
        contractTX.addTag('Content-Type', 'application/json');
        yield arweave.transactions.sign(contractTX, wallet);
        const response = yield arweave.transactions.post(contractTX);
        if (response.status === 200 || response.status === 208) {
            return contractTX.id;
        }
        else {
            throw new Error('Unable to write Contract Initial State');
        }
    });
}
exports.createContractFromTx = createContractFromTx;
