// arlocal/bin/graphql/resolvers.js
"use strict";
var __awaiter = (this && this.__awaiter) || function (thisArg, _arguments, P, generator) {
    function adopt(value) { return value instanceof P ? value : new P(function (resolve) { resolve(value); }); }
    return new (P || (P = Promise))(function (resolve, reject) {
        function fulfilled(value) { try { step(generator.next(value)); } catch (e) { reject(e); } }
        function rejected(value) { try { step(generator["throw"](value)); } catch (e) { reject(e); } }
        function step(result) { result.done ? resolve(result.value) : adopt(result.value).then(fulfilled, rejected); }
        step((generator = generator.apply(thisArg, _arguments || [])).next());
    });
};
var __importDefault = (this && this.__importDefault) || function (mod) {
    return (mod && mod.__esModule) ? mod : { "default": mod };
};
Object.defineProperty(exports, "__esModule", { value: true });
exports.parseCursor = exports.encodeCursor = exports.newCursor = exports.resolvers = void 0;
const moment_1 = __importDefault(require("moment"));
const encoding_1 = require("../utils/encoding");
const query_1 = require("./query");
const ar_1 = require("../utils/ar");
const DEFAULT_PAGE_SIZE = 10;
const MAX_PAGE_SIZE = 100;
const fieldMap = {
    id: 'transactions.id',
    anchor: 'transactions.last_tx',
    recipient: 'transactions.target',
    tags: 'transactions.tags',
    fee: 'transactions.reward',
    quantity: 'transactions.quantity',
    data_size: 'transactions.data_size',
    data_type: 'transactions.content_type',
    parent: 'transactions.parent',
    owner: 'transactions.owner',
    owner_address: 'transactions.owner_address',
    signature: 'transactions.signature',
    block_id: 'blocks.id',
    block_timestamp: 'blocks.mined_at',
    block_height: 'blocks.height',
    block_previous: 'blocks.previous_block',
};
const blockFieldMap = {
    id: 'blocks.id',
    timestamp: 'blocks.mined_at',
    height: 'blocks.height',
    previous: 'blocks.previous_block',
    extended: 'blocks.extended',
};
exports.resolvers = {
    Query: {
        transaction: (_, queryParams, { connection }) => __awaiter(void 0, void 0, void 0, function* () {
            const params = {
                id: queryParams.id,
                blocks: true,
                select: fieldMap,
            };
            const result = yield (0, query_1.generateQuery)(params, connection);
            return result[0];
        }),
        transactions: (_, queryParams, { connection }) => __awaiter(void 0, void 0, void 0, function* () {
            var _a, _b;
            const { timestamp, offset } = (0, exports.parseCursor)(queryParams.after || (0, exports.newCursor)());
            const pageSize = Math.min(queryParams.first || DEFAULT_PAGE_SIZE, MAX_PAGE_SIZE);
            const params = {
                limit: pageSize + 1,
                offset,
                ids: queryParams.ids || undefined,
                to: queryParams.recipients || undefined,
                from: queryParams.owners || undefined,
                tags: queryParams.tags || undefined,
                blocks: true,
                since: timestamp,
                select: fieldMap,
                minHeight: ((_a = queryParams.block) === null || _a === void 0 ? void 0 : _a.min) || undefined,
                maxHeight: ((_b = queryParams.block) === null || _b === void 0 ? void 0 : _b.max) || undefined,
                sortOrder: queryParams.sort || undefined,
            };
            const results = (yield (0, query_1.generateQuery)(params, connection));
            const hasNextPage = results.length > pageSize;
            return {
                pageInfo: {
                    hasNextPage,
                },
                edges: () => __awaiter(void 0, void 0, void 0, function* () {
                    return results.slice(0, pageSize).map((result, index) => {
                        return {
                            cursor: (0, exports.encodeCursor)({ timestamp, offset: offset + index + 1 }),
                            node: result,
                        };
                    });
                }),
            };
        }),
        block: (_, queryParams, { connection }) => __awaiter(void 0, void 0, void 0, function* () {
            if (queryParams.id) {
                return (yield (0, query_1.generateBlockQuery)({
                    select: blockFieldMap,
                    id: queryParams.id,
                }, connection))[0];
            }
            else {
                return null;
            }
        }),
        blocks: (_, queryParams, { connection }) => __awaiter(void 0, void 0, void 0, function* () {
            const { timestamp, offset } = (0, exports.parseCursor)(queryParams.after || (0, exports.newCursor)());
            const pageSize = Math.min(queryParams.first || DEFAULT_PAGE_SIZE, MAX_PAGE_SIZE);
            let ids = [];
            let minHeight = 0;
            let maxHeight = MAX_PAGE_SIZE;
            if (queryParams.ids) {
                ids = queryParams.ids;
            }
            if (queryParams.height && queryParams.height.min) {
                minHeight = queryParams.height.min;
            }
            if (queryParams.height && queryParams.height.max) {
                maxHeight = queryParams.height.max;
            }
            const query = (0, query_1.generateBlockQuery)({
                ids,
                select: blockFieldMap,
                minHeight,
                maxHeight,
                sortOrder: queryParams.sort || 'HEIGHT_ASC',
                limit: pageSize + 1,
                offset,
                before: timestamp,
            }, connection);
            const results = yield query;
            // @ts-ignore
            const hasNextPage = results.length > pageSize;
            return {
                pageInfo: {
                    hasNextPage,
                },
                edges: () => __awaiter(void 0, void 0, void 0, function* () {
                    // @ts-ignore
                    return results.slice(0, pageSize).map((result, index) => {
                        return {
                            cursor: (0, exports.encodeCursor)({ timestamp, offset: offset + index + 1 }),
                            node: result,
                        };
                    });
                }),
            };
        }),
    },
    Transaction: {
        tags: (parent) => {
            if (typeof parent.tags === 'string')
                parent.tags = JSON.parse(parent.tags);
            return parent.tags.map(encoding_1.utf8DecodeTag);
        },
        recipient: (parent) => {
            return parent.recipient.trim();
        },
        data: (parent) => {
            return {
                size: parent.data_size || 0,
                type: parent.data_type,
            };
        },
        quantity: (parent) => {
            return {
                ar: (0, ar_1.winstonToAr)(parent.quantity || 0),
                winston: parent.quantity || 0,
            };
        },
        fee: (parent) => {
            return {
                ar: (0, ar_1.winstonToAr)(parent.fee || 0),
                winston: parent.fee || 0,
            };
        },
        block: (parent) => {
            if (parent.block_id) {
                return {
                    id: parent.block_id,
                    previous: parent.block_previous,
                    timestamp: parent.block_timestamp,
                    height: parent.block_height,
                };
            }
        },
        owner: (parent) => {
            return {
                address: parent.owner_address,
                key: parent.owner,
            };
        },
        parent: (parent) => {
            if (parent.parent) {
                return {
                    id: parent.parent,
                };
            }
        },
    },
    Block: {
        /*
        reward: (parent) => {
          return {
            address: parent.extended.reward_addr,
            pool: parent.extended.reward_pool,
          };
        },
        size: (parent) => {
          return parent.extended?.block_size;
        },
        */
        timestamp: (parent) => {
            return (0, moment_1.default)(parent === null || parent === void 0 ? void 0 : parent.timestamp).unix();
        },
    },
};
const newCursor = () => (0, exports.encodeCursor)({ timestamp: (0, moment_1.default)().toISOString(), offset: 0 });
exports.newCursor = newCursor;
const encodeCursor = ({ timestamp, offset }) => {
    const str = JSON.stringify([timestamp, offset]);
    return Buffer.from(str).toString('base64');
};
exports.encodeCursor = encodeCursor;
const parseCursor = (cursor) => {
    try {
        const [timestamp, offset] = JSON.parse(Buffer.from(cursor, 'base64').toString());
        return { timestamp, offset };
    }
    catch (error) {
        throw new Error('invalid cursor');
    }
};
exports.parseCursor = parseCursor;
